hibernate-core/src/main/java/org/hibernate/query/sqm/tree/domain/SqmPolymorphicRootDescriptor.java

```diff
@@ -37,56 +36,67 @@
 import org.hibernate.query.sqm.SqmPathSource;
 import org.hibernate.type.descriptor.java.JavaType;
 
+import static java.util.Collections.unmodifiableMap;
+
 /**
- * Acts as the EntityValuedNavigable for a "polymorphic query" grouping
+ * Acts as the {@link EntityDomainType} for a "polymorphic query" grouping.
  *
  * @author Steve Ebersole
  */
 public class SqmPolymorphicRootDescriptor<T> implements EntityDomainType<T> {
-	private final Set<EntityDomainType<?>> implementors;
-	private final Map<String, PersistentAttribute> commonAttributes;
+	private final Set<EntityDomainType<? extends T>> implementors;
+	private final Map<String, PersistentAttribute<? super T,?>> commonAttributes;
 
 	private final JavaType<T> polymorphicJavaType;
 
 	public SqmPolymorphicRootDescriptor(
 			JavaType<T> polymorphicJavaType,
-			Set<EntityDomainType<?>> implementors) {
+			Set<EntityDomainType<? extends T>> implementors) {
 		this.polymorphicJavaType = polymorphicJavaType;
-
 		this.implementors = implementors;
+		this.commonAttributes = unmodifiableMap( inferCommonAttributes( implementors ) );
+	}
 
-		final Map<String, PersistentAttribute> workMap = new HashMap<>();
-
-		final ArrayList<EntityDomainType<?>> implementorsList = new ArrayList<>( implementors );
-
+	/**
+	 * The attributes of a "polymorphic" root are the attributes which are
+	 * common to all subtypes of the root type.
+	 */
+	private Map<String, PersistentAttribute<? super T, ?>> inferCommonAttributes(Set<EntityDomainType<? extends T>> implementors) {
```

Comment:
Yes, that's true, but in context it's (fuzzily) OK, because what we're basically doing is inferring that if the field is on every subentity it's because it's actually declared higher up the type hierarchy, i.e. by the "polymorphic root" or one of its super types, for which we don't actually have a metamodel. So in principle the typing is correct, it's just the *validation* that's not totally rigorous. (But that lack of validation was a pre-existing problem.)